sql/src/main/java/io/crate/analyze/expressions/ExpressionAnalyzer.java

```diff
@@ -613,8 +613,10 @@ public Symbol visitArrayComparisonExpression(ArrayComparisonExpression node, Exp
             Symbol leftSymbol = process(node.getLeft(), context);
             Symbol arraySymbol = process(node.getRight(), context);
 
+            if (arraySymbol.valueType().equals(DataTypes.UNDEFINED)) {
+                arraySymbol = cast(arraySymbol, new ArrayType(leftSymbol.valueType()));
+            }
             DataType arraySymbolType = arraySymbol.valueType();
-
             if (!DataTypes.isCollectionType(arraySymbolType)) {
                 throw new IllegalArgumentException(
                     SymbolFormatter.format("invalid array expression: '%s'", arraySymbol));
```

Comment:
If we removed or deferred this check (and the object below), until after the function allocation, it works fine.